File: worker/src/worker/reconciler.py
"""Startup + reconnect reconciliation.

MinIO's bucket notifications are best-effort — events can be missed
across container restarts, network blips, or short outages. We sweep
Mongo on startup and after every reconnect, picking up:

  * rows still in `pending` (never processed)
  * rows stuck in `processing` past `stuck_after_minutes` (worker
    crashed mid-docling)

For each, we reset `processing` → `pending` and run the same pipeline
the listener would have. Single-flight so a flood of uploads on
startup doesn't pin the worker.
"""

from __future__ import annotations

import asyncio
import logging
from typing import Awaitable, Callable

from pymongo.collection import Collection

from . import store

log = logging.getLogger(__name__)


ProcessOne = Callable[[store.UploadRecord], Awaitable[None]]


async def run_once(
    col: Collection,
    *,
    process_one: ProcessOne,
    stuck_after_minutes: int,
) -> int:
    """Process all pending/stuck rows once. Returns the count handled."""

    rows = store.list_pending_or_stuck(col, stuck_after_minutes=stuck_after_minutes)
    if not rows:
        return 0
    log.info("reconciler found rows", count=len(rows))
    handled = 0
    for row in rows:
        # Reclaim stuck `processing` rows back to `pending` so
        # `process_one` can claim them with the same atomic transition
        # the live listener uses.
        store.reset_to_pending(col, row.mongo_id)
        try:
            await process_one(row)
            handled += 1
        except Exception:
            log.exception(
                "reconciler failed to process row",
                key=row.key,
            )
    return handled

File: worker-rs/src/reconciler.rs
//! Startup + reconnect reconciliation.
//!
//! MinIO bucket notifications are best-effort — events can be missed
//! across restarts, network blips, or short outages. We sweep Mongo on
//! startup and after every reconnect, picking up:
//!
//!  * rows still in `pending` (never processed)
//!  * rows stuck in `processing` past `stuck_after` (worker crashed
//!    mid-docling)
//!
//! For each, we reset `processing` → `pending` and run the same pipeline
//! the listener would have.

use std::time::Duration;

use bson::Document;
use mongodb::Collection;

use crate::mongo;

pub async fn run_once<F, Fut>(
    coll: &Collection<Document>,
    stuck_after: Duration,
    process: F,
) -> usize
where
    F: Fn(mongo::UploadRecord) -> Fut,
    Fut: std::future::Future<Output = ()>,
{
    let rows = match mongo::list_pending_or_stuck(coll, stuck_after).await {
        Ok(r) => r,
        Err(e) => {
            tracing::error!(error = %e, "reconciler list failed");
            return 0;
        }
    };

    if rows.is_empty() {
        return 0;
    }
    tracing::info!(count = rows.len(), "reconciler picked up rows");
    let mut handled = 0;
    for row in rows {
        let id = row.mongo_id.clone();
        let coll_reset = coll.clone();
        let _ = mongo::reset_to_pending(&coll_reset, &id).await;
        process(row).await;
        handled += 1;
    }
    handled
}
